Migration from decompress

AlbSed opened the pull request:
### Decompression functionality enhancements:
* Added support for `.zlib` files by implementing a new `unpack_zlib` function and registering it as a supported format in the `Decompressor` class. [[1]](diffhunk://#diff-2fe5258b5bde88f5008ce627d43b6bfa5673eacb6c54c5fdb8d1264671b58869L5-R67) [[2]](diffhunk://#diff-2fe5258b5bde88f5008ce627d43b6bfa5673eacb6c54c5fdb8d1264671b58869R114)
* Enhanced error handling in the `run` method of `src/component.py` by introducing a `graceful` mode, which logs errors instead of raising exceptions when enabled.

### Dependency and project configuration updates:
* Updated the `Dockerfile` to use Python 3.13 and integrated `uv` for dependency management.
* Added a `pyproject.toml` file to define project metadata and dependencies, including new libraries such as `py7zr` and `python-snappy`.

### CI/CD pipeline improvements:
* Updated the GitHub Actions workflow (`.github/workflows/push.yml`) to include steps for pushing and tagging the Decompress processor in the Developer Portal. [[1]](diffhunk://#diff-f3fc934cf0d89bdf07de358896ff90f1202585df812cf606206d1830a9949811R23-R28) [[2]](diffhunk://#diff-f3fc934cf0d89bdf07de358896ff90f1202585df812cf606206d1830a9949811R202-R212) [[3]](diffhunk://#diff-f3fc934cf0d89bdf07de358896ff90f1202585df812cf606206d1830a9949811R233-R241)

### Documentation updates:
* Updated the `README.md` to include `.zlib` as a supported file format.

Commit:
progress -> (zlib & snappy implementation), edited tests requirements, Dockerfile optimization

Commit:
uv implementation, Dockerfile update, zlib implementation

Commit:
Pipeline setup, README and uv update, component added graceful settings

soustruh reviewed (approved):
I see no problems in the updated code, but there seems to be a bug in the original code you didn't touch which probably prevents .tar.gz files to work as expected. 👀

soustruh commented on src/decompress/decompress.py:
```diff
@@ -2,51 +2,69 @@
-SUPPORTED_FORMATS = [
-    ".7z", ".tar.bz2", ".tbz2", ".gz", ".tar.gz", ".tgz", ".tar", ".tar.xz", ".txz", ".zip"
-]
+SUPPORTED_FORMATS = [".7z", ".tar.bz2", ".tbz2", ".gz", ".zlib", ".tar.gz", ".tgz", ".tar", ".tar.xz", ".txz", ".zip"]
```
OK, it actually works for tar.gz, as I missed the `.gz` item here, as one cannot find it quickly in the list when the list is inline and sorted. 🤦‍♂️ 😁

Please add trailing comma so ruff splits this list to one line per item and sort the lines alphabetically (maybe with the slightest optimalization exception of `.zip` being first as it's probably the most common format users will pass into the processor) 🙏

soustruh replied:
OK, it actually works for tar.gz, as I missed the .gz item here at first, as one cannot find it quickly in the list when the list is inline and unsorted. 🤦‍♂️ 😁

Please add trailing comma so ruff splits this list to one line per item and sort the items alphabetically (maybe with the slightest optimalization exception of .zip being first as it's probably the most common format users will pass into the processor) 🙏

soustruh commented on src/component.py:
```diff
@@ -44,13 +46,21 @@ def run(self):
                 file_out_path = self._get_out_path(file)
-                d.decompress(file, file_out_path)
+                try:
+                    d.run_decompressor(file, file_out_path)
```
Please remove the `file_extension in SUPPORTED_FORMATS` here as in the worst case scenario, this check (using a list, so with linear complexity) is done 3 times, I simply don't understand why anyone would write such a code.

My suggestion is to remove the logic from `__init__` and replace it with a single call to `any … endswith` down in the run_decompressor method. The method might return True for success, False for skipped and raise an exception in case of failure, or it might stay stateless and simply log the skipped files on its own.

soustruh reviewed (changes requested):
I see no problems in the updated code, but there seems to be a bug in the original code you didn't touch which probably prevents .tar.gz files to work as expected. 👀

EDIT: OK, it's not a bug, but please make it a bit prettier and easier to read. Thanks! 🙇

Commit:
simplified logic for format checks, removed unused imports and dependencies

Commit:
revert to original tests setup

Commit:
tests fix

AlbSed commented:
@soustruh look again please. I adjusted it based on what we talked about and I fixed the unittest expectations

```
(processor-unzip) C:\Users\alber\DATA\_work\processor-unzip>python -m unittest discover
.compare_source_and_expected (datadirtest.datadirtest.TestDataDir.compare_source_and_expected)
to_folder_false ... Running C:/Users/alber/DATA/_work/processor-unzip/src/component.py with configuration from C:\Users\alber\AppData\Local\Temp\test_data
Extraction starting.
C:\Users\alber\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\shutil.py:1281: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tarobj.extractall(extract_dir, filter=filter)
Extraction done.
Tests passed successfully
ok
compare_source_and_expected (datadirtest.datadirtest.TestDataDir.compare_source_and_expected)
to_folder_true ... Running C:/Users/alber/DATA/_work/processor-unzip/src/component.py with configuration from C:\Users\alber\AppData\Local\Temp\test_data
Extraction starting.
Extraction starting.
Extraction done.
Extraction done.
Tests passed successfully
Tests passed successfully
ok

----------------------------------------------------------------------
Ran 2 tests in 0.450s

OK
.
----------------------------------------------------------------------
Ran 2 tests in 0.479s

OK
```

soustruh commented on src/decompress/decompress.py:
```diff
@@ -1,52 +1,69 @@
+        shutil.register_unpack_format("zlib", [".zlib"], unpack_zlib)
 
-    def decompress(self, file_path, file_out_path) -> None:
+    def run_decompressor(self, file_path, file_out_path) -> bool | tuple[bool, str]:
```
OK, I know I kinda suggested modification of the way this method works, but please please do not return values in this way mixing Python and Go coding style. Returning tuples OR single values, more so when the docstring simply says `Returns: None` is a really bad practice 😿

soustruh commented on src/component.py:
```diff
@@ -44,15 +46,22 @@ def run(self):
+                    raise UserException(f"Error processing file {file}: {e}")
+
+            if not done:
+                logging.warning(f"Error processing file {file}: {done[1]}")
```
This will definitely cause a crash 👀 if not done means the returned value is None, False, 0, "", [] or (), but definitely not two-item tuple, so accessing done[1] will definitely fail 🙁

soustruh commented on src/decompress/decompress.py:
```diff
@@ -1,52 +1,69 @@
-SUPPORTED_FORMATS = [
-    ".7z", ".tar.bz2", ".tbz2", ".gz", ".tar.gz", ".tgz", ".tar", ".tar.xz", ".txz", ".zip"
-]
+SUPPORTED_FORMATS = [".zip", ".7z", ".gz", ".tar", ".tar.bz2", ".tar.gz", ".tar.xz", ".tbz2", ".tgz", ".txz", ".zlib"]
```
Please really put a comma after ".zlib" and then run ruff and see what happens 🙏

soustruh reviewed (changes requested):
I'd be able to accept passing tuple under certain circumstances, like when the logic is really simple and when unpacking the values, the results are stored in a well-named variables, but receiving simply `done` and not knowing that it can be bool or tuple, so might or might not have multiple items, that is just a bad practice.

If anything, you can find inspiration in the Go language, which usually returns ok, err instead of using exceptions (there are none in fact in the language), maybe always return something like `status: bool, message: str`, but I still think my original suggestion was the cleanest solution – pass everything to the run_decompress method, if it returns True, do nothing, if it returns False (unsupported format, log unsupported format) and if it encounters an exception, you can handle it right inside the method (it can receive the graceful argument too). And yep, as you would handle logging and exceptions inside in such case, there might be no need to return anything at all and warning about unsupported files also inside the function. This will also help you simplify the logic of the higher level method, making it more readable, and do stuff in the lower level method… 💡

Commit:
changes to simplify the components logic and for moving its processing parts to Decompress class

Commit:
reduce 'graceful' logic only to error with broken archive

Commit:
After research I switched back to graceful logic on all. Updated README

soustruh reviewed (approved):
Everything seems OK now 👍
